mlflow/pyfunc/__init__.py

```diff
@@ -266,7 +270,163 @@ def _load_model_env(path):
+    ordering declared in schema if necessary. Any extra columns are ignored.
+
+    For column types, we make sure the types match schema or can be safely converted to match the
+    input schema.
+    """
+    if isinstance(pdf, list):
+        pdf = pandas.DataFrame(pdf)
+    if not isinstance(pdf, pandas.DataFrame):
+        message = 'Expected input to be DataFrame or list. Found: %s' % type(pdf).__name__
+        raise MlflowException(message)
+
+    if input_schema.has_column_names():
+        # make sure there are no missing columns
+        col_names = input_schema.column_names()
+        expected_names = set(col_names)
+        actual_names = set(pdf.columns)
+        missing_cols = expected_names - actual_names
+        extra_cols = actual_names - expected_names
+        # Preserve order from the original columns, since missing/extra columns are likely to
+        # be in same order.
+        missing_cols = [c for c in col_names if c in missing_cols]
+        extra_cols = [c for c in pdf.columns if c in extra_cols]
+        if missing_cols:
+            message = ("Model input is missing columns {0}."
+                       " Note that there were extra columns: {1}".format(missing_cols, extra_cols))
+            raise MlflowException(message)
+    else:
+        # The model signature does not specify column names => we can only verify column count.
+        if len(pdf.columns) < len(input_schema.columns):
+            message = ("Model input is missing input columns. The model signature declares "
+                       "{0} input columns but the provided input only has "
+                       "{1} columns. Note: the columns were not named in the signature so we can "
+                       "only verify their count.").format(len(input_schema.columns),
+                                                          len(pdf.columns))
+            raise MlflowException(message)
+        col_names = pdf.columns[:len(input_schema.columns)]
+    col_types = input_schema.column_types()
+    new_pdf = pandas.DataFrame()
+    for i, x in enumerate(col_names):
+        new_pdf[x] = _enforce_type(x, pdf[x], col_types[i])
+    return new_pdf
+
+
+PyFuncOutput = Union[pandas.DataFrame, pandas.Series, np.ndarray, list]
+
+
+class PyFuncModel(object):
+    """
+    MLflow 'python function' model.
+
+    Wrapper around model implementation and metadata. This class is not meant to be constructed
+    directly. Instead, instances of this class are constructed and returned from
+    py:func:`mlflow.pyfunc.load_model`.
+
+    ``model_impl`` can be any Python object that implements the `Pyfunc interface
+    <https://mlflow.org/docs/latest/python_api/mlflow.pyfunc.html#pyfunc-inference-api>`_, and is
+    by invoking the model's ``loader_module``.
+
+    ``model_meta`` contains model metadata loaded from the MLmodel file.
+    """
+
+    def __init__(self, model_meta: Model, model_impl: Any):
```

Comment:
Is `model_meta` optional? It's a required param but nobody checks until `metadata()` is called. Looks like an avoidable runtime error?

Reply:
Not a major issue. I should have added a nit to this.